### PR TITLE
MultiprocessFileCache for efficient and easy cache with pytorch dataloader (num_workers>0)

### PR DESCRIPTION
**Pull request by belltailjp**
In case we combine pfio (+HDFS) with PyTorch DataLoader (num_workers>0), due to the following
reasons that are complexly related each other, I noticed that it is hard to make efficient/effective use of pfio FileCache easily.
* With HDFS, setting num_workers>0 is necessary for practical data loading performance
* The current pfio FileCache supports therad-safety but cannot be process-safe
* Currently PyTorch DataLoader is designed to destroy all the worker processes at the end of each epoch

https://github.com/pytorch/pytorch/pull/35795 would solve part of the problems, but being difficult for efficiently setting higher `num_workers` will remain unsolved.

After serious considerations, I figured out that implementing multi-process safe file cache is the simplest way to overcome this issue.
This PR is the introduction of the `MultiprocessFileCache` which imitates the interface and behavior of the existing `FileCache`.

(* modified from the initial implementation)
The new `MultiprocessFileCache` is supposed to be initialized in a single master process and then distributed to its child processes through pickle.
All of these cloned object handle the same cache file(s) created by the master process. The `MultiprocessFileCache` does not keep file descriptors or file objects of the cache files as in the current `FileCache`, but instead it only keeps the filenames. Every time it reads/writes the cache it opens them. This provides process safety.
Cache files remain undeleted as long as the `MultiprocessFileCache` in the *master* process alives, no matter if `MultiprocessFileCache` is destroyed earlier in the *child* process by explicit call of `close`, `exit` nor GC. When the object is destroyed in the master, caches are made sure to be destroyed. This behavior is necessary to prevent the cache is lost in every epoch, when using with PyTorch DataLoader.

A micro-banchmark showed approximately 2x speedup (loading ImageNet zipped val images (including jpeg decode) with `num_workers=16`, cpu=24 setting) in my internal environment.

### CODE REVIEW

**Comment by belltailjp** [comment]
I have made a few modifications based on an internal discussion
- Use a file for offset and lengths cache (as done in the current FileCache)
- Support what files to use for cache
- Fixed exclusive access control bug
As the result, it allows an intra-node shared cache file (sometimes useful e.g., in imbalanced distributed training setup), and cache file persistency.

In a distributed ImageNet training performance (ResNet50/fp16/8xV100/bs=32 per worker), I confirmed approximately 12% speedup in average, which is very significant.

**Comment by belldandyxtq** [comment]
Is https://github.com/pfnet/pfio/blob/d37b16f55f87fc9217d22e6b0b23bee94b07e8b8/pfio/cache/file_cache.py#L285 helpful in this case? Or you want to also speed up the first epoch data loading

**Comment by belldandyxtq** [comment]
Can you please merge to the latest master? the test fails due to #126 and is fixed by #127

**Comment by belltailjp** [comment]
I think it's indeed very helpful feature.
In the current implementation, the same thing can be achieved by `MultiprocessFileCache(..., index_cache_file='xxxx', data_cache_file='yyy', cleanup=False)`.
`{index|data}_cache_file` arguments are intended to support an intra-node shared cache file, but by combining with `cleanup=False` the caches remain persisted.
If it's better to have the same interface (also better to control timing of persistency process), I think I can implement it, but something equivalent to `{index|data}_cache_file` should still be needed in the initializer for its original purpose, which may be a bit confusing if we have both.
\# Oh, btw it doesn't support what is equivalent to `FileCache.preload`..

**Comment by belltailjp** [comment]
Hi, it's been quite a while, but I update the MultiprocessFileCache implementation. I think it's now ready for review.

* Similar `preserve` and `preload` interface to the existing FileCache.
* Test

File: pfio/cache/multiprocess_file_cache.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+            else:
+                raise ose
+
+    def _put(self, i, data):
+        if self.closed:
+            return
+        assert 0 <= i < self.length
+
+        offset = self.buflen * i
+        index_fd = os.open(self.index_file.name, os.O_RDWR)
```
Comment (belldandyxtq) [inline pfio/cache/multiprocess_file_cache.py:188]: ditto

File: pfio/cache/multiprocess_file_cache.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+        '''
+
+        index_file = os.path.join(self.dir, '{}.cachei'.format(name))
+        data_file = os.path.join(self.dir, '{}.cached'.format(name))
+
+        if any(os.path.exists(p) for p in (index_file, data_file)):
+            raise ValueError('Specified cache name "{}" already exists in {}'
+                             .format(name, self.dir))
+
+        fd = os.open(self.index_file.name, os.O_RDONLY)
```
Comment (belldandyxtq) [inline pfio/cache/multiprocess_file_cache.py:277]: ditto

File: pfio/cache/multiprocess_file_cache.py [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+        else:
+            self.dir = dir
+        os.makedirs(self.dir, exist_ok=True)
+
+        self.closed = False
+        self._frozen = False
+        self._master_pid = os.getpid()
+        self.data_file = _NoOpenNamedTemporaryFile(self.dir, self._master_pid)
+        self.index_file = _NoOpenNamedTemporaryFile(self.dir, self._master_pid)
+        index_fd = os.open(self.index_file.name, os.O_RDWR)
```
Comment (belldandyxtq) [inline pfio/cache/multiprocess_file_cache.py:105]: Do we close this?

**Review by belldandyxtq — CHANGES_REQUESTED** [review]
Thanks again for your great contribution on this very interesting PR. I have the following questions:
1. Can you please provide some examples on how this cache can be properly used with the `Pytorch DataLoader`? It can be in the docstring.

2. If I understand correctly, the reason why we do not keep the `fd` is to pickle to the subprocesses. But I am concerned that opening/closing the file for every single read usually significantly damage the I/O performance. Is there anyway to reduce the amount of opening/closing? I am not very sure about how exactly the `Pytorch DataLoader` works inside, but is it possible to do things like this:
1. close the fds in master after all the `put`, and set to None
2. fork
3. check the fds in the subprocesses, if it is None, open the fd
In this way we only need one read per subprocesses.

File: pfio/cache/multiprocess_file_cache.py [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+    Arguments:
+        length (int): Length of the cache array.
+
+        do_pickle (bool):
+            Do automatic pickle and unpickle inside the cache.
+
+        dir (str): The path to the directory to place cache data in
+            case home directory is not backed by fast storage device.
+
```
[the remark this answers is not in the document]
Comment (belltailjp) [reply pfio/cache/multiprocess_file_cache.py:82]: It's not actually documented in `FileCache`. Should they be both documented?

**Comment by belltailjp** [comment]
> Can you please provide some examples on how this cache can be properly used with the Pytorch DataLoader? It can be in the docstring.

I added it to the `MultiprocessFileCache` docstring.

> If I understand correctly, the reason why we do not keep the fd is to pickle to the subprocesses. But I am concerned that opening/closing the file for every single read usually significantly damage the I/O performance. Is there anyway to reduce the amount of opening/closing? I am not very sure about how exactly the Pytorch DataLoader works inside, but is it possible to do things like this:

Thanks for the advice! Indeed, it is possible to make sure to open and close the fds only once in each process. I applied this change (though I actually saw no performance improvement in a micro-benchmark).   
Also, dangling fd problem pointed out by your comments are resolved along together.

File: pfio/cache/multiprocess_file_cache.py (hunk 3 above)
Comment (belltailjp) [reply pfio/cache/multiprocess_file_cache.py:151]: All done in the recent fix :+1:

**Review by belldandyxtq — CHANGES_REQUESTED** [review]

File: pfio/cache/multiprocess_file_cache.py [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+        '''
+
+        index_file = os.path.join(self.dir, '{}.cachei'.format(name))
+        data_file = os.path.join(self.dir, '{}.cached'.format(name))
+
+        if any(os.path.exists(p) for p in (index_file, data_file)):
+            raise ValueError('Specified cache name "{}" already exists in {}'
+                             .format(name, self.dir))
+
```
[the remark this answers is not in the document]
Comment (belltailjp) [reply pfio/cache/multiprocess_file_cache.py:354]: Right! I forgot it, now I pushed the fix along with a new test case.

**Review by belldandyxtq — APPROVED** [review]

**Pull request merged**